File: app/src/main/java/com/tokyonth/installer/activity/SettingsActivity.java
package com.tokyonth.installer.activity;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.View;
import android.widget.TextView;

import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;
import com.kyleduo.switchbutton.SwitchButton;
import com.tokyonth.installer.Constants;
import com.tokyonth.installer.R;
import com.tokyonth.installer.adapter.SettingsAdapter;
import com.tokyonth.installer.utils.StatusBarColorUtils;
import com.tokyonth.installer.widget.CustomizeDialog;
import com.tokyonth.installer.bean.SettingsBean;
import com.tokyonth.installer.utils.GetAppInfoUtils;
import com.tokyonth.installer.utils.FileUtils;
import com.tokyonth.installer.utils.SPUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.Objects;

public class SettingsActivity extends AppCompatActivity {

    private SwitchButton switchButtonUseSysPkg;
    private TextView textViewSysPkgName;
    private TextView textViewApkCache;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        StatusBarColorUtils.setStatusBarDarkIcon(this,
                !(boolean) SPUtils.getData(Constants.SP_NIGHT_MODE, false));
        setContentView(R.layout.activity_settings);
        initViewData();
        initSettings();
    }

    private void initViewData() {
        ArrayList<SettingsBean> settingsBeanArrayList = new ArrayList<>();
        settingsBeanArrayList.add(new SettingsBean(getString(R.string.title_show_perm),
                getString(R.string.summary_show_perm),
                R.drawable.ic_verified_user_24px, getResources().getColor(R.color.color0)));
        settingsBeanArrayList.add(new SettingsBean(getString(R.string.title_show_act),
                getString(R.string.summary_show_act),
                R.drawable.ic_widgets_24px, getResources().getColor(R.color.color1)));
        settingsBeanArrayList.add(new SettingsBean(getString(R.string.vibrate),
                getString(R.string.install_vibrate),
                R.drawable.ic_waves_24px, getResources().getColor(R.color.color2)));

        SettingsAdapter adapter = new SettingsAdapter(settingsBeanArrayList);
        RecyclerView rvSettings = findViewById(R.id.rv_settings_item);
        rvSettings.setLayoutManager(new LinearLayoutManager(this));
        rvSettings.setAdapter(adapter);
        adapter.setOnItemClick((view, pos, bool) -> {
            switch (pos) {
                case 0:
                    SPUtils.putData(Constants.SP_SHOW_PERM, bool);
                    break;
                case 1:
                    SPUtils.putData(Constants.SP_SHOW_ACT, bool);
                    break;
                case 2:
                    SPUtils.putData(Constants.SP_VIBRATE, bool);
                    break;
            }
        });
        TextView textViewVersion = findViewById(R.id.tv_version);
        textViewVersion.append(GetAppInfoUtils.getVersionName(this));

        textViewApkCache = findViewById(R.id.tv_apk_cache);
        textViewSysPkgName = findViewById(R.id.tv_pkg_name);
        switchButtonUseSysPkg = findViewById(R.id.cb_use_sys_pkg);
        switchButtonUseSysPkg.setOnCheckedChangeListener((compoundButton, isChecked) -> SPUtils.putData(Constants.SP_USE_SYS_PKG, isChecked));

        String cacheSize = FileUtils.byteToString(FileUtils.getFileOrFolderSize(new File(Constants.CACHE_APK_DIR)));
        textViewApkCache.setText(getString(R.string.text_apk_cache, cacheSize));

        findViewById(R.id.card_apk_cache).setOnClickListener(view -> {
            FileUtils.deleteFolderFile(Constants.CACHE_APK_DIR, true);
            Snackbar.make(findViewById(R.id.coordinator_layout), getString(R.string.text_apk_cache_complete), Snackbar.LENGTH_SHORT).show();
            textViewApkCache.setText(getString(R.string.text_apk_cache, cacheSize));
        });

        findViewById(R.id.card_pkg).setOnClickListener(view -> {
            View inView = View.inflate(SettingsActivity.this, R.layout.layout_input_pkg, null);
            final TextInputEditText edit = inView.findViewById(R.id.et_sys_pkg_name);

            CustomizeDialog.getInstance(this)
                    .setTitle(R.string.text_title_input)
                    .setView(inView)
                    .setPositiveButton(R.string.dialog_ok, (dialog, which) -> {
                        String str = Objects.requireNonNull(edit.getText()).toString().trim();
                        if (str.isEmpty()) {
                            Snackbar.make(findViewById(R.id.coordinator_layout), getString(R.string.text_input_empty), Snackbar.LENGTH_SHORT).show();
                        } else {
                            SPUtils.putData(Constants.SYS_PKG_NAME, str);
                            textViewSysPkgName.setText(str);
                        }
                    })
                    .setNegativeButton(R.string.dialog_btn_cancel, null)
                    .setCancelable(false).create().show();
        });
    }

    private void initSettings() {
        boolean useSysPkg = (boolean) SPUtils.getData(Constants.SP_USE_SYS_PKG, false);
        switchButtonUseSysPkg.setChecked(useSysPkg);
        if (!SPUtils.getData(Constants.SYS_PKG_NAME, Constants.SYS_PKG_NAME).equals(Constants.SYS_PKG_NAME)) {
            textViewSysPkgName.setText(SPUtils.getData(Constants.SYS_PKG_NAME, Constants.SYS_PKG_NAME).toString());
        }
    }

}

File: app/src/main/java/com/tokyonth/installer/activity/SettingsActivity.kt
package com.tokyonth.installer.activity

import android.content.pm.PackageManager
import android.view.LayoutInflater
import android.view.MenuItem
import android.widget.CompoundButton
import androidx.appcompat.app.AppCompatDelegate
import androidx.recyclerview.widget.LinearLayoutManager
import com.catchingnow.icebox.sdk_client.IceBox
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.Snackbar
import com.tokyonth.installer.BuildConfig
import com.tokyonth.installer.R
import com.tokyonth.installer.adapter.SettingsAdapter
import com.tokyonth.installer.adapter.SettingsAdapter.*
import com.tokyonth.installer.data.SPDataManager
import com.tokyonth.installer.databinding.ActivitySettingsBinding
import com.tokyonth.installer.databinding.LayoutInputPkgBinding
import com.tokyonth.installer.utils.NotificationUtils
import com.tokyonth.installer.utils.PackageUtils
import com.tokyonth.installer.utils.ktx.*
import rikka.shizuku.Shizuku

class SettingsActivity : BaseActivity() {

    private val binding: ActivitySettingsBinding by lazyBind()

    private var settingsAdapter = SettingsAdapter()

    private var shizukuPermission: Shizuku.OnRequestPermissionResultListener? = null

    override fun setBinding() = binding

    override fun initView() {
        setSupportActionBar(binding.toolBar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
        }
        initSettings()
        initViewStatus()
    }

    override fun initData() {
        shizukuPermission = Shizuku.OnRequestPermissionResultListener { requestCode, grantResult ->
            if (grantResult == PackageManager.PERMISSION_GRANTED) {
                updateShizuku(requestCode)
            }
        }
        Shizuku.addRequestPermissionResultListener(shizukuPermission!!)
        permissionHelper?.registerCallback { all, code ->
            if (all) {
                updateShizuku(code)
            } else {
                val str = when (code) {
                    1 -> string(R.string.shizuku_permission_request)
                    2 -> string(R.string.icebox_permission_request)
                    else -> ""
                }
                Snackbar.make(binding.root, str, Snackbar.LENGTH_SHORT).show()
            }
        }
    }

    private fun initSettings() {
        val local = SPDataManager.instance
        settingsAdapter.apply {
            setOnItemActionListener(object : OnItemActionListener {
                override fun onSwitch(view: CompoundButton, position: Int, bool: Boolean) {
                    when (position) {
                        0 -> local.setShowPermission(bool)
                        1 -> local.setShowActivity(bool)
                        2 -> {
                            val b = NotificationUtils.checkNotification(this@SettingsActivity)
                            if (b) {
                                local.setDefaultSilent(bool)
                            } else {
                                view.isChecked = false
                            }
                        }
                        3 -> local.setAutoDel(bool)
                        4 -> {
                            if (bool) {
                                AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
                            }
                            local.setFollowSystem(bool)
                        }
                    }
                }

                override fun onClick(position: Int) {
                    if (position == 5) {
                        MaterialAlertDialogBuilder(this@SettingsActivity)
                            .setSingleChoiceItems(
                                R.array.install_mode_arr,
                                SPDataManager.instance.getInstallMode()
                            ) { dialog, which ->
                                when (which) {
                                    0 -> local.setInstallMode(which)
                                    1 -> shizukuCheck(which)
                                    2 -> iceBoxCheck(which)
                                }
                                settingsAdapter.updateInstallMode()
                                dialog.dismiss()
                            }
                            .setNegativeButton(R.string.dialog_btn_cancel, null)
                            .show()
                    }
                }
            })
        }
        binding.rvSettings.apply {
            layoutManager = LinearLayoutManager(this@SettingsActivity)
            adapter = settingsAdapter
        }
    }

    private fun initViewStatus() {
        val local = SPDataManager.instance
        binding.cbUseSysPkg.isChecked = local.isUseSystemPkg()
        binding.tvPkgName.text = local.getSystemPkg()

        val cacheSize = externalCacheDir?.fileOrFolderSize()?.toMemorySize() ?: 0
        binding.tvApkCache.text = string(R.string.text_apk_cache, cacheSize)
        binding.tvAppVersion.text = string(R.string.text_settings_version, BuildConfig.VERSION_NAME)
        binding.cbUseSysPkg.setOnCheckedChangeListener { _, isChecked ->
            local.setUseSystemPkg(isChecked)
        }
        binding.cardPkg.click {
            val pkgBinding =
                LayoutInputPkgBinding.inflate(LayoutInflater.from(this@SettingsActivity))
            MaterialAlertDialogBuilder(this@SettingsActivity)
                .setTitle(R.string.text_title_input)
                .setView(pkgBinding.root)
                .setPositiveButton(R.string.dialog_btn_ok) { _, _ ->
                    val name = pkgBinding.etSystemPkg.text.toString()
                    if (name.isEmpty()) {
                        Snackbar.make(
                            binding.root,
                            getString(R.string.text_input_empty),
                            Snackbar.LENGTH_SHORT
                        ).show()
                    } else {
                        local.setSystemPkg(name.trim())
                        binding.tvPkgName.text = name.trim()
                    }
                }
                .setNegativeButton(R.string.dialog_btn_cancel, null)
                .show()
        }
        binding.cardCache.click {
            externalCacheDir?.deleteFolderFile(true)
            Snackbar.make(
                binding.root,
                string(R.string.clean_complete),
                Snackbar.LENGTH_SHORT
            ).show()
            binding.tvApkCache.text = string(R.string.text_apk_cache, "0B")
        }
    }

    private fun iceBoxCheck(requestCode: Int) {
        if (!PackageUtils.isIceBoxClientAvailable(this)) {
            Snackbar.make(
                binding.root,
                getString(R.string.not_install_icebox),
                Snackbar.LENGTH_SHORT
            ).show()
        } else {
            permissionHelper?.start(arrayOf(IceBox.SDK_PERMISSION), requestCode)
        }
    }

    private fun shizukuCheck(requestCode: Int) {
        if (!PackageUtils.isShizukuClientAvailable(this)) {
            Snackbar.make(
                binding.root,
                getString(R.string.shizuku_not_installed),
                Snackbar.LENGTH_SHORT
            ).show()
        } else {
            val p = checkShizukuPermission(requestCode)
            if (p) {
                updateShizuku(requestCode)
            }
            //permissionHelper?.start(arrayOf(ShizukuProvider.PERMISSION), requestCode)
        }
    }

    private fun updateShizuku(code: Int) {
        SPDataManager.instance.setInstallMode(code)
        settingsAdapter.updateInstallMode()
    }

    private fun checkShizukuPermission(code: Int): Boolean {
        if (Shizuku.isPreV11()) {
            return false
        }
        return try {
            if (Shizuku.checkSelfPermission() == PackageManager.PERMISSION_GRANTED) {
                true
            } else if (Shizuku.shouldShowRequestPermissionRationale()) {
                false
            } else {
                Shizuku.requestPermission(code)
                false
            }
        } catch (e: Exception) {
            Snackbar.make(
                binding.root,
                getString(R.string.shizuku_not_run),
                Snackbar.LENGTH_SHORT
            ).show()
            false
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        Shizuku.removeRequestPermissionResultListener(shizukuPermission!!)
        permissionHelper?.dispose()
    }

}
